rel/app/standalone.exs

```diff
@@ -0,0 +1,102 @@
+defmodule Standalone do
+  @moduledoc false
+  require Logger
+
+  @doc """
+  Copies ERTS into the release.
+  """
+  @spec copy_erlang(Mix.Release.t()) :: Mix.Release.t()
+  def copy_erlang(release) do
+    {erts_source, erts_bin_dir, erts_lib_dir, _erts_version} = erts_data()
+
+    erts_destination_source = Path.join(release.path, "vendor/bin")
+    File.mkdir_p!(erts_destination_source)
+
+    erts_source
+    |> Path.join("bin")
+    |> File.cp_r!(erts_destination_source, fn _, _ -> false end)
+
+    _ = File.rm(Path.join(erts_destination_source, "erl"))
+    _ = File.rm(Path.join(erts_destination_source, "erl.ini"))
+
+    erts_destination_source
+    |> Path.join("erl")
+    |> File.write!(~S"""
+    #!/bin/sh
+    SELF=$(readlink "$0" || true)
+    if [ -z "$SELF" ]; then SELF="$0"; fi
+    BINDIR="$(cd "$(dirname "$SELF")" && pwd -P)"
+    ROOTDIR="${ERL_ROOTDIR:-"$(dirname "$(dirname "$BINDIR")")"}"
+    EMU=beam
+    PROGNAME=$(echo "$0" | sed 's/.*\///')
+    export EMU
+    export ROOTDIR
+    export BINDIR
+    export PROGNAME
+    exec "$BINDIR/erlexec" ${1+"$@"}
+    """)
+    executable!(Path.join(erts_destination_source, "erl"))
+
+    # Copy lib
+    erts_destination_lib = Path.join(release.path, "lib")
+    File.mkdir_p!(erts_destination_lib)
+
+    erts_lib_dir
+    |> File.cp_r!(erts_destination_lib, fn _, _ -> false end)
+
+    # copy *.boot files to <resource_path>/bin
+    erts_destination_bin =  Path.join(release.path, "bin")
+
+    boot_files = erts_bin_dir |> Path.join("*.boot") |> Path.wildcard() |> Enum.map(& String.split(&1, "/") |> List.last())
+
+    File.mkdir_p!(erts_destination_bin)
+
+    for boot_file <- boot_files do
+      erts_bin_dir |> Path.join(boot_file) |> File.cp!(Path.join(erts_destination_bin, boot_file))
+    end
+
+    %{release | erts_source: erts_source}
+  end
+
+  @doc """
+  Copies elixir into the release.
+  """
+  @spec copy_elixir(Mix.Release.t(), elixir_version :: String.t()) :: Mix.Release.t()
+  def copy_elixir(release, elixir_version) do
+    # download and unzip
+    standalone_destination = Path.join(release.path, "vendor/elixir")
+    download_elixir_at_destination(standalone_destination, elixir_version)
```

Comment:
Note that we're already downloading Elixir in out bootstrap script https://github.com/livebook-dev/livebook/blob/main/.github/scripts/app/bootstrap_mac.sh#L123:L124 so we don't need to do it again. I think it's better if this function really just copies the currently running Elixir, similarly how `copy_erlang` does that for OTP. This way it's going to be super easy to test it with Elixir main when developing too.

Reply:
I am worried though that we will have even less control over how Elixir is installed. For example, some package installers like to change the `bin` files. :'(

The downside of using the Elixir precompiled though is that it uses the minimum OTP version and if you are building from scratch then it is guaranteed you compiled it to the latest.

Reply:
That's fair. But we'll always be building Livebook.app with our bootstrap script where we have full control so I don't think that's gonna be an issue in practice. We need to be running the bootstrap script because we need to have full control over OTP for exactly same reason.